streams/src/test/java/org/apache/kafka/streams/EqualityCheck.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams;
+
+public final class EqualityCheck {
```

Comment:
This allows us to verify all the properties we wish to hold w.r.t. equals/hashCode.

Most of them are required by either the Java spec or by common sense:
https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#equals(java.lang.Object)
https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#hashCode()

The one "extra" requirement is that inequal objects should have different hashcodes. The spec specifically says this is allowed, but it will cause poor map/set performance, since they will be guaranteed to collide and wind up creating linked lists.
Since we're only validating our own implementations, it's ok to insist on this property.